File: lib/queue.go
package lib

import (
	"bytes"
	"encoding/binary"
	"strconv"

	"github.com/samuel/go-zookeeper/zk"
)

type Queue struct {
	*SyncPrimitive
}

func NewQueue(address, root string) (*Queue, error) {
	sp, err := NewSyncPrimitive(address)
	if err != nil {
		return nil, err
	}
	q := &Queue{SyncPrimitive: sp}
	q.root = root
	exists, _, err := q.conn.Exists(root)
	if err == nil && !exists {
		_, _ = q.conn.Create(root, []byte{}, 0, zk.WorldACL(zk.PermAll))
	}
	return q, nil
}

func (q *Queue) Produce(i int) error {
	buf := new(bytes.Buffer)
	_ = binary.Write(buf, binary.BigEndian, int32(i))
	_, err := q.conn.Create(
		q.root+"/element",
		buf.Bytes(),
		zk.FlagSequence,
		zk.WorldACL(zk.PermAll),
	)
	return err
}

func (q *Queue) Consume() (int, error) {
	for {
		children, _, ch, err := q.conn.ChildrenW(q.root)
		if err != nil {
			return -1, err
		}
		if len(children) == 0 {
			<-ch
			continue
		}
		minNode := children[0]
		minVal, _ := strconv.Atoi(minNode[len("element"):])
		for _, c := range children {
			v, _ := strconv.Atoi(c[len("element"):])
			if v < minVal {
				minVal = v
				minNode = c
			}
		}
		data, stat, err := q.conn.Get(q.root + "/" + minNode)
		if err != nil {
			continue
		}
		err = q.conn.Delete(q.root+"/"+minNode, stat.Version)
		if err != nil {
			continue
		}
		var val int32
		_ = binary.Read(bytes.NewReader(data), binary.BigEndian, &val)
		return int(val), nil
	}
}

File: lib/queue.py
import sys
import time
from lib.sync_primitive import SyncPrimitive

class Queue(SyncPrimitive):
    def __init__(self, address, root):
        super().__init__(address)  # Inicializa a conexão com o ZooKeeper
        self.root = root  # Caminho do nó raiz da fila
        self._ensure_root()  # Garante que o nó raiz existe

    def _ensure_root(self):
        self.zk.ensure_path(self.root)  # Cria o nó raiz se ele não existir

    def produce(self, value):
        # Adiciona um elemento à fila criando um nó sequencial
        self.zk.create(f"{self.root}/element-", str(value).encode(), sequence=True)

    def consume(self):
        while True:
            children = self.zk.get_children(self.root)  # Obtém os nós filhos (elementos da fila)
            if not children:
                time.sleep(1)  # Se a fila estiver vazia, aguarda 1 segundo
                continue
            sorted_children = sorted(children)  # Ordena os nós filhos
            first_child = sorted_children[0]  # Pega o nó mais antigo (menor sequência)
            data, _ = self.zk.get(f"{self.root}/{first_child}")  # Obtém o valor do nó
            self.zk.delete(f"{self.root}/{first_child}")  # Remove o nó da fila
            return int(data.decode())  # Retorna o valor consumido